### PR TITLE
feat: support writing only internal messages to WAL

### PR DESCRIPTION
**Pull request by rach-id**
## Description

Closes https://github.com/celestiaorg/celestia-core/issues/1247

This PR adds a configuration flag to limit the WAL to only contain internal messages.

This is useful because, as perceived in experiments, external_messages_WAL_writes ~= number_of_validators * internal_messages_WAL_writes, it reduces significantly the number of writes on the WAL. This makes sense because, by default, the WAL contains: votes, block parts, proposals, etc, from all the peers. In the meantime what we really need is the validator's own messages, and especially signature, not to double sign (To be confirmed because I guess, unless there is some bug in the signing process or some weird uncommon issue, the signature should be the same for each round).

~The configuration is added as experimental and is turned off by default.~

The configuration is enabled by default.

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use
  [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by rootulp — DISMISSED** [review]

File: config/config.go [hunk 1]
```diff
@@ -941,8 +941,12 @@ func (cfg *FastSyncConfig) ValidateBasic() error {
 // including timeouts and details about the WAL and the block structure.
 type ConsensusConfig struct {
 	RootDir string `mapstructure:"home"`
-	WalPath string `mapstructure:"wal_file"`
-	walFile string // overrides WalPath if set
+	// Experimental: If set to true, only internal messages will be written
+	// to the WAL. External messages like votes, proposals
+	// block parts, will not be written
+	OnlyInternalWal bool   `mapstructure:"only_internal_wal"`
```
Comment (evan-forbes) [inline config/config.go:947]: nit:

we might want to be more explicit with the default value being false

File: config/toml.go [hunk 2]
```diff
@@ -456,6 +456,12 @@ version = "{{ .FastSync.Version }}"
 #######################################################
 [consensus]
 
+# Experimental: if set to "true", only internal messages will be
+# written to the WAL. External messages like votes, proposal,
+# block parts, will not be written.
+# Use at your own risk!
```
Comment (evan-forbes) [inline config/toml.go:462]: can we make an issue to remind us to delete the "Use at your own risk?" after we are confident that its working as intended? either that or we should remove it now and be very confident that its safe before we actually include it in any release candidate.

**Review by evan-forbes — COMMENTED** [review]
I think the only blocking thing is that we should change e2e test to not use the internal wal

if it breaks, then we might want to reserve such a change to a separate branch only for experimentation purposes until we have ironed out all the fixes.

as a side note for posterity, writing the wal here is async, but it still has a profound impact on delay because when we write internal messages we flush the wal. meaning that all the msgs that build up in the buffer here must be written to disc before we are able to proceed.

**Comment by rach-id** [comment]
> that we should change e2e test to not use the internal wal

You mean we should make the default true and see if the e2e tests fail? or better add case in e2e where only internal wal is true?

**Comment by evan-forbes** [comment]
> You mean we should make the default true and see if the e2e tests fail? or better add case in e2e where only internal wal is true?

any mechanism that changes it in the e2e test, particularly for nodes that restart. As you mentioned, we could change the default to true, we could also only change this value in the e2e tests.

**Review by staheri14 — DISMISSED** [review]

**Comment by rach-id** [comment]
Making it a draft so that we don't accidentally merge it. I will be making it ready for review once I run the right tests

File: config/config.go (hunk 1 above)
Comment (rach-id) [reply config/config.go:948]: done

File: config/toml.go (hunk 2 above)
Comment (rach-id) [reply config/toml.go]: Removed from the PR and made the default be `true`

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**